docs: add/modify specs to describe auction types

jeonghoyeo7 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

This is to add the details of specs for the module.

closes: #84

## Tasks

- [x] Draft 01_concept.md
- [x] Draft 02_state.md
- [x] Draft 03_state_transitions.md
- [x] Draft 04_messages.md 
- [x] Draft 05_end_block.md
- [x] Modifying related .proto files and related refactoring



## References
- Continued work from https://github.com/tendermint/fundraising/issues/81

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://go.dev/blog/godoc).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

jeonghoyeo7 committed:
docs: update specs

jeonghoyeo7 committed:
docs: update specs

jeonghoyeo7 committed:
docs: update specs including testing

jaybxyz commented on proto/fundraising/fundraising.proto:
```diff
@@ -94,22 +94,25 @@ message BaseAuction {
   ];
 
+  // current number of winning bidders
+  uint64 number_winning_bidders = 13;
```
How should we handle the value for `FixedPriceAuction`? Should we increment by one when a bid gets success?

jeonghoyeo7 replied:
Yes it can be one way.

jaybxyz commented on proto/fundraising/fundraising.proto:
```diff
@@ -121,27 +124,21 @@ message FixedPriceAuction {
-// of coin and bidders bid to purchase the amounts of coin. It creates
-// competition for the price, not how fast bidders can bid.
-message EnglishAuction {
+// BatAuction defines a batch auction type. It allows each bidder
+// to participate in the auction by placing limit orders with a bid price
+// chosen freely at any time within the auction period. An order book is created
+// to record the bids with various bid prices.
```
A little typo in `BatAuction` and I add more context around the `BatchAuction`. 
Feel free to add/update the suggested lines.
```suggestion
// BatchAuction defines a batch auction type. It allows bidders to participate in the auction 
// by placing their limit orders with a bid price they are willing to bid within the auction period.
// They can place multiple bids with different bid prices and if they want to modify their existing bid, 
// they only have an option to modify with a higher bid price. Under the hood, an order book is created 
// to record the bids to calculate winners.
```

jeonghoyeo7 replied:
I will reflect this.

jaybxyz commented on proto/fundraising/fundraising.proto:
```diff
@@ -121,27 +124,21 @@ message FixedPriceAuction {
-  // extend_rate specifies the rate that decides if the auction needs another
-  // round
-  string extend_rate = 4 [
+  // rate_extended_round specifies the rate that decides if the auction needs
+  // another round
+  string rate_extended_round = 3 [
```
I think the name should be `extended_round_rate`.

jeonghoyeo7 replied:
I will reflect this.

jaybxyz commented on x/fundraising/spec/02_state.md:
```diff
@@ -46,7 +46,12 @@ type AuctionI interface {
 	SetVestingSchedules([]VestingSchedule) error
-
+	
+	GetWinningPrice() sdk.Dec  // To load WinningPrice at the current time
+	GetNumberWinningBidders() uint64
+	
+	GetRemainingSellingCoin() sdk.Coin // To calculate RemainingSellingCoin
```
Is there a reason why you don't have Setter functions?

jeonghoyeo7 replied:
I will add the setters.

jaybxyz reviewed (changes requested):
I think we need to update `tx.proto` as well. Also, I see a number of places where `EnglishAuction` is used in comments and docs.

jeonghoyeo7 committed:
docs: update specs including testing

jeonghoyeo7 commented:
- I've changed based on the comments.
 - Also changed the remaining "English" related.
 - "sequence" --> "BidId", "Eligible" --> "isWInner"

But, it seems it needs to add AllowedBidders to MsgCreate...Auction in msg.go

jaybxyz commented:
>But, it seems it needs to add AllowedBidders to MsgCreate...Auction in msg.go


All auctions are closed when they are created. The fundraising module is designed to delegate the control to an external module.

jaybxyz commented on proto/fundraising/fundraising.proto:
```diff
@@ -239,11 +242,11 @@ message Bid {
 
   // id specifies an index of a bid for the bidder
-  uint64 id = 3;
+  uint64 bid_id = 3;
```
Since id is already inside `Bid` struct, I think it can just be `id`.

jeonghoyeo7 replied:
I updated based on the comments 1,2,3 above.
For adding NumWinningBidders to CreateFixedPriceAuction of keeper/auction.go, I added NumWinningBidders = 0.

jaybxyz reviewed (changes requested):
I have a couple things that are not covered in this PR.

1. `NewBaseAuction` needs to have an added argument(s). Update the affected codebase.
     - `NumberWinningBidders` 
2.  We can follow SDK convention of using shorten word for `Number`, so I would change from `NumberWinningBidders` to `NumWinningBidders` (It is shorten for a number of winning bidders)
3. I found a number of places that are using `sequence`. They need to be updated.

jeonghoyeo7 committed:
docs: update specs including testing

jaybxyz commented:
Once you merge from the latest `main` branch and resolve conflicts, I will approve this PR.

jeonghoyeo7 committed:
docs: update specs including testing
